xls/synthesis/timing_characterization_client_main.py

```diff
@@ -54,85 +88,304 @@ def _synth(stub: synthesis_service_pb2_grpc.SynthesisServiceStub,
   return stub.Compile(request)
 
 
-def _run_binop(
-    op: str, kop: str, stub: synthesis_service_pb2_grpc.SynthesisServiceStub
-) -> List[delay_model_pb2.DataPoint]:
-  """Characterize a binary operation via synthesis server.
+def _run_unary_bitwise(
+    kop: str, model: delay_model_pb2.DelayModel,
+    stub: synthesis_service_pb2_grpc.SynthesisServiceStub) -> None:
+  # Add op_model to protobuf message
+  add_op_model = model.op_models.add(op=kop)
+  add_op_model.estimator.regression.factors.add(
+    source=delay_model_pb2.DelayFactor.OPERAND_BIT_COUNT,
+    operand_number=0)
+
+  op = ENUM2NAME_MAP[kop]
+
+  # Compute samples
+  results = []
+  for bit_count in [1] + list(range(2, MAX_WIDTH, 2)):
+    op_type = f'bits[{bit_count}]'
+    ir_text = op_module_generator.generate_ir_package(op, op_type, (op_type,))
+    module_name = f'{op}_{bit_count}'
+    mod_generator_result = op_module_generator.generate_verilog_module(module_name, ir_text)
+    top_name = module_name + '_wrapper'
+    verilog_text = op_module_generator.generate_parallel_module([mod_generator_result], top_name)
+    try:
+      result = _synth(stub, verilog_text, top_name)
+    except grpc.RpcError:
+      break
+
+    ps = 1e12/result.max_frequency_hz
+    result = delay_model_pb2.DataPoint()
+    result.operation.op = kop
+    result.operation.bit_count = bit_count
+    operand = result.operation.operands.add()
+    operand.bit_count = bit_count
+    result.delay = int(ps)
+    results.append(result)
```

Comment:
The bit of code from line 108 to 122 is repeated a number of times, how about making that a separate function, something like:

def synthesize_ir(
  ir_text: str,
  op: str,
  result_bit_count: int,
  operand_bit_counts = Sequence[int])
  -> delay_model_pb2.DataPoint:
 ...

Reply:
(solved)